File: DOTNET_CSharp/Tests/ArraySearchTests.cs
using Library;
using Xunit;

namespace Tests {
    public class ArraySearchTests {
        [Theory]
        
        [InlineData(new[] { 1, 2, 3, 4, 5 }, 3, 2)]
        [InlineData(new[] { 1, 2, 3, 4, 5 }, 1, 0)]
        [InlineData(new[] { 1, 2, 3, 4, 5 }, 5, 4)]
        [InlineData(new[] { 1, 2, 3, 4, 5 }, 6, 5)]
        [InlineData(new[] { 1, 2, 3, 4, 5 }, 0, -1)]
        [InlineData(new[] { 1, 2, 4, 4, 5 }, 3, 2)]
        [InlineData(new[] { 1, 2, 4, 5 }, 4, 2)]
        public void SearchTest(int[] input, int item, int expected) {
            // Act
            var actual = input.SearchIndex(item, 0, input.Length - 1);

            // Assert
            Assert.Equal(expected, actual);
        }
    }
}

File: FSharp/Tests/ArraySearchTests.fs
namespace Tests

module ArraySerchTests =
    open Xunit
    open Library.ArraySearch
    open Library.Utils

    let data = [|
        // Item in the Middle
        [| ~~[ 1 .. 5 ]; ~~3; ~~2 |]

        // First Item
        [| ~~[ 1 .. 5 ]; ~~1; ~~0 |]

        // Last Item
        [| ~~[ 1 .. 5 ]; ~~5; ~~4 |]
        
        // Non-Existent Item From Right
        [| ~~[ 1 .. 5 ]; ~~6; ~~5 |]

        // Non-Existent Item from Left
        [| ~~[ 1 .. 5 ]; ~~0; ~~(-1) |]
        
        // Non-Existent Item in Between
        [| ~~[ 1 .. 5 ]; ~~3; ~~2 |]
        
        // Repeated Item
        [| ~~[ 1; 2; 4; 4; 5 ]; ~~4; ~~2 |]
    |]

    [<Theory>]
    [<MemberData("data")>]
    let Test (ar, item, expected) = 
        let actual = searchIndex ar item 0 (ar.Length - 1)
        Assert.Equal(expected, actual)
